src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1248,6 +1220,123 @@ private void ParseModifiers(SyntaxListBuilder tokens, bool forAccessors)
             }
         }
 
+        private bool IsRefStruct(out bool isRefFollowedByStructOrPartialStruct, out bool isTypeDeclaration)
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.RefKeyword);
+
+            var nextToken = PeekToken(1);
+
+            isRefFollowedByStructOrPartialStruct = nextToken.Kind == SyntaxKind.StructKeyword ||
+                (nextToken.ContextualKind == SyntaxKind.PartialKeyword && this.PeekToken(2).Kind == SyntaxKind.StructKeyword);
+
+            if (isRefFollowedByStructOrPartialStruct)
+            {
+                isTypeDeclaration = true;
+                return true;
+            }
+
+            isTypeDeclaration = IsTypeDeclaration(out nextToken);
+            return isTypeDeclaration && nextToken.Kind == SyntaxKind.StructKeyword;
+        }
+
+        private static bool CanFollowTypeName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.DotToken:                   // partial.     qualified
+                case SyntaxKind.ColonColonToken:            // partial::    alias
+                case SyntaxKind.LessThanToken:              // partial<     generic type
+                case SyntaxKind.OpenBracketToken:           // partial[     array
+                    return true;
+            }
+
+            return false;
+        }
+
+        private static bool CanFollowMemberName(SyntaxKind kind)
+        {
+            switch (kind)
+            {
+                case SyntaxKind.LessThanToken:              // partial<     generic method
+                case SyntaxKind.OpenBraceToken:             // partial {    property
+                case SyntaxKind.EqualsGreaterThanToken:     // partial =>   property
+                case SyntaxKind.OpenParenToken:             // partial(     method
+                    return true;
+            }
+
+            return false;
+        }
+
+        private enum PartialModFlags
+        {
+            PartialType,
+            PartialMember,
+            TreatAsModifier,
+            NotModifier
+        }
+
+        private PartialModFlags ShouldPartialBeTreatedAsModifier()
+        {
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.PartialKeyword);
+
+            int peekIndex = 1;
+            SyntaxToken currentToken;
+
+            while (GetModifier(currentToken = PeekToken(peekIndex)) != DeclarationModifiers.None)
+            {
+                if (SyntaxFacts.IsContextualKeyword(currentToken.ContextualKind))
+                {
+                    var nextTokenKind = PeekToken(peekIndex + 1).Kind;
+                    if (nextTokenKind == SyntaxKind.ThisKeyword ||
+                        (nextTokenKind == SyntaxKind.IdentifierToken
+                             ? CanFollowMemberName(PeekToken(peekIndex + 2).Kind)
+                             : CanFollowTypeName(nextTokenKind)))
```

Comment:
I'm not sure if this even worth it, `ShouldAsyncBeTreatedAsModifier` just expects a single additional partial mod. @CyrusNajmabadi might have a better idea here.

Reply:
these complex expressions really hurt readability for me.  I don't find dense expressoins a virtue in the parser *except* when constructing node.  For complex case checking, i would far prefer one check after another, with each check graciously stating what cases it is considering.

Reply:
```CanFollowTypeName(nextTokenKind)```

How do you know you're after a type?

Reply:
The point is to determine if the contextual keyword is not a modifier. It could be either a type name or a member name, so we look at the the next token to see if it can actually follow a type name or member name.

Reply:
> these complex expressions really hurt readability for me. I don't find dense expressoins a virtue in the parser except when constructing node. For complex case checking, i would far prefer one check after another, with each check graciously stating what cases it is considering.

:)

Reply:
I read that. I just replied to your last comment to clarify the intent of those helper methods.

Reply:
I thikn it would be more sensible to invert things.  Instead of htis logic of "can htis follow".  say you had hte list of modifiers.  You simply go to each contextual modifiers and ask "shoudl this be treated like a modifier".  And the answer is usually pretty simple.

1. if it's followed by a real modifier, it should be treated as a modifier.    "partial public".  "partial" is def a modifier.
2. If it's *not* the last, or second-to-last modifier, it's a modifier.      

Then all you have to check is what to do if it's the last or second-to-last modifier.  Note: if it's second-to-last, that means it's followed by another contextual keyword (otherwise rule '1' would have kicked in).  

So now, you just need to check what follows.  There are the simple cases.  like "class/struct/interface/namespace/etc.".  These make them into modifiers.  

Then you have to do the cases where those identifiers might be the return-type/member name.  But you just check for those, and handle accordingly.  

This puts all the logic for modifiers in a single location, and gives a single spot where we're trying to handle all the different cases for where contextual modifiers should be treated as identifiers and not modifiers.

And, importantly, it gives full context.  i.e. we can know if we're in the actual hard case, versus cases '1' or '2' and we can just trivially enumerate what are hte interesting things we need to check for.